pkg/openfeature/event_executor.go

```diff
@@ -0,0 +1,319 @@
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	v, ok := provider.(EventHandler)
+	if !ok {
+		return nil
+	}
+
+	oldProvider := e.namedProviderReference[associatedClient]
+
+	// register shutdown semaphore for new named provider
+	sem := make(chan interface{})
+
+	newProvider := &providerReference{
+		eventHandler:          &v,
+		clientNameAssociation: associatedClient,
+		metadata:              provider.Metadata(),
+		shutdownSemaphore:     sem,
+	}
+
+	e.namedProviderReference[associatedClient] = newProvider
+	return e.listenAndShutdown(newProvider, oldProvider)
+}
+
+// listenAndShutdown is a helper to start concurrent listening to new provider events and invoke shutdown hook of the
+// old provider
+func (e *eventExecutor) listenAndShutdown(newProvider *providerReference, oldReference *providerReference) error {
+	go func(newProvider *providerReference, oldReference *providerReference) {
+		for {
+			select {
+			case event := <-(*newProvider.eventHandler).EventChannel():
+				err := e.triggerEvent(event, newProvider.clientNameAssociation, newProvider.isDefault)
+				if err != nil {
+					e.logger.Error(err, fmt.Sprintf("error handling event type: %s", event.EventType))
+				}
+			case <-newProvider.shutdownSemaphore:
+				return
+			}
+		}
+	}(newProvider, oldReference)
+
+	// shutdown old provider handling
+	if oldReference == nil {
+		return nil
+	}
+
+	select {
+	case oldReference.shutdownSemaphore <- "":
+		return nil
+	case <-time.After(200 * time.Millisecond):
+		return fmt.Errorf("event handler timeout waiting for handler shutdown")
+	}
+}
+
+// triggerEvent performs the actual event handling
+func (e *eventExecutor) triggerEvent(event Event, clientNameAssociation string, isDefault bool) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	group, gCtx := errgroup.WithContext(ctx)
+
+	group.Go(func() error {
+		// first run API handlers
+		for _, c := range e.apiRegistry[event.EventType] {
+			e.executeHandler(*c, event)
+		}
+
+		// then run Client handlers for name association
+
+		// first direct associates
+		associateClientRegistry := e.scopedRegistry[clientNameAssociation]
+		for _, c := range associateClientRegistry.callbacks[event.EventType] {
+			e.executeHandler(*c, event)
+		}
+
+		if !isDefault {
+			return nil
+		}
+
+		// handling the default provider - invoke default provider bound handlers by filtering
+
+		var defaultHandlers []EventCallback
+
+		for clientName, registry := range e.scopedRegistry {
+			if _, ok := e.namedProviderReference[clientName]; !ok {
+				defaultHandlers = append(defaultHandlers, registry.callbacks[event.EventType]...)
+			}
+		}
+
+		for _, c := range defaultHandlers {
+			e.executeHandler(*c, event)
+		}
+
+		return nil
+	})
+
+	// wait for completion or timeout
+	select {
+	case <-time.After(handlerExecutionTime):
```

Comment:
This is not spec'd, and honestly it could be changed later, but I think a per-handler timeout would be more useful.

Reply:
I removed the timeout handling as this is not specified in the spec. further, seems other sdk implementations do not follow a similar approach, which make go-sdk to function differently